sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/scopes.scala

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.catalyst.analysis.{Resolver, UnresolvedAttribute}
+import org.apache.spark.sql.catalyst.plans.logical.{CoGroup, LogicalPlan, MapGroups}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.types.DataType
+
+
+/**
+ * An expression that has to be resolved against a scope of resolved attributes.
+ */
+case class ScopedExpression(expr: Expression, scope: Seq[Attribute])
```

Comment:
I'm a bit worried about rule order issue here. How can we guarantee that we resolve `expr` with the scoped attributes? `expr` is a normal child of `ScopedExpression` and any rule can resolve it?

Reply:
If we removed `exprs` from `override def children`, then no rule can resolve `exprs`, right?

Reply:
For instance, for a groupBy.agg, the scope is the data attributes of the groupBy, which are not required to be resolved at that stage.